packages/components/src/border-control/styles.ts

```diff
@@ -154,12 +176,22 @@ export const borderWidthControl = () => css`
 	/* Target the InputControl's backdrop */
 	&&& ${ BackdropUI } {
 		border: none;
+		${ rtl( { borderTopLeftRadius: 0, borderBottomLeftRadius: 0 } )() }
 	}
 
 	/* Specificity required to overcome UnitControl padding */
 	/* See packages/components/src/unit-control/styles/unit-control-styles.ts */
 	&&& input {
 		${ rtl( { paddingRight: 0 } )() }
+
+		/*
+		 * Allows the input control's backdrop to overlay its box shadow over
+		 * the BorderControl's wrapper border.
+		 */
+		&:focus ~ ${ BackdropUI } {
+			box-shadow: 0 0 0 ${ CONFIG.borderWidthFocus }
```

Comment:
Should these box-shadows be set to inset instead, for consistency with the color indicator?

<img width="148" alt="image" src="https://user-images.githubusercontent.com/14988353/167528209-b7daf7c7-2ebb-40f6-93e1-c0086614e1e7.png">

From clicking around in the editor sidebar, it looks like most of the other input fields currently use inset for the focus styling.

Reply:
> Should these box-shadows be set to inset instead 

They weren't inset so that the box-shadow overlayed the border from the overall `BorderControl`. That is, no "double border" as mentioned in the original issue. 

I chose that as opposed to removing borders from the wrapper and managing additional partial borders and edge cases for all the internal components.

> for consistency with the color indicator?

The color indicator's focus style should already be consistent i.e. it isn't `inset` either. It appears from that screenshot that you are highlighting the dropdown button's hover style?

<img width="496" alt="Screen Shot 2022-05-10 at 1 06 30 pm" src="https://user-images.githubusercontent.com/60436221/167534710-12e90b73-8251-4f7b-bd01-ba1f700e9e13.png">

From the Storybook, here's how it looks to me when I toggle on focus for both the button and the input field.

<img width="296" alt="Screen Shot 2022-05-10 at 1 04 02 pm" src="https://user-images.githubusercontent.com/60436221/167534492-3404fce9-4fc6-49ce-b45b-da0233faab02.png">

I did wonder about making the hover style match the focus but thought I'd see what came up there in review. So that is still something I could address in a follow-up.

> From clicking around in the editor sidebar, it looks like most of the other input fields currently use inset for the focus styling.

I might need to re-test via the block editor and see if any style conflicts are occurring there. Thanks!

Reply:
Thanks for the extra context!

> It appears from that screenshot that you are highlighting the dropdown button's hover style?

Ah, yes, I was, that's it. A better comparison might be `InputControl` / `UnitControl` in fields that we render near the Border controls, (e.g. Block Spacing or Border Radius):

| Border focused | Border Radius focused |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/167538442-92d317f5-aec8-4923-8e07-e65096ffb614.png) | ![image](https://user-images.githubusercontent.com/14988353/167538456-78bd6dcf-1194-43dc-accf-9edc162c91c9.png) |

I _think_ the latter's style comes from the following line:
https://github.com/WordPress/gutenberg/blob/be92ce8d06f1e282e9cfb401e43790d60ef8f137/packages/components/src/input-control/styles/input-control-styles.tsx#L301

With the UnitControl in BorderRadius, it looks like that one currently has a thicker "looking" inset style that avoids the double-border and looks a bit closer to the design in #40897 to my eyes.

For InputControl, it appears it currently achieves a 2px effective border in [backdropFocusedStyles](https://github.com/WordPress/gutenberg/blob/be92ce8d06f1e282e9cfb401e43790d60ef8f137/packages/components/src/input-control/styles/input-control-styles.tsx#L292) by doing the following:

* Sets a `1px` border with the color set to the border focus color when the component is focused
* Sets a `1px` inset box shadow with the color set to the border focus color

That way it winds up looking effectively like a 2px border, without increasing the apparent size of the element. Not to say that we need to do that necessarily for this component, too, but just thought I'd add that in here in case it helps!

> I did wonder about making the hover style match the focus but thought I'd see what came up there in review. So that is still something I could address in a follow-up.

Tackling other inconsistencies in follow-ups sounds good to me! I'm sure @shaunandrews will have a better idea of some of the trade-offs of these fine-tuning decisions than me, also 😀

Happy to do more testing, too 🙂

Reply:
I've taken another look at the block editor and controls within the sidebar. I'm struggling to find any consistency as to whether the controls are concerned with focus styling not extending the bounds of the non-focused control dimensions.

#### Without inset box-shadow

- Toggle controls
- Buttons (General buttons, menu item checkboxes, tabs, dropdown buttons, panel headers)
- Font Size Picker (ToggleGroup view)
- CustomSelectControl
- TextControl (both input and textarea)
- Links
- CheckboxControl
- RangeControl drag handles

#### With inset box-shadow

- InputControl
- UnitControl
- SelectControl

Given the widespread use of UnitControl, that sort of makes the balance more even. 

As the current approach will address the "double borders" issue, I'm inclined to proceed with it, perhaps just with a tweak to make the border control dropdown button's hover style match its focus style.

What do you think @andrewserong?

Reply:
Thank you for the extra digging and testing.

> For InputControl, it appears it currently achieves a 2px effective border in [backdropFocusedStyles](https://github.com/WordPress/gutenberg/blob/be92ce8d06f1e282e9cfb401e43790d60ef8f137/packages/components/src/input-control/styles/input-control-styles.tsx#L292) by doing the following:

I will take another look at getting the resulting width consistent (i.e. `2px`). In my testing the box-shadow sets a width of `COLORS.ui.borderFocus` which shows as `1.5px` and renders as `2px`?

> Sets a 1px border with the color set to the border focus color when the component is focused

While making these changes this appeared to make the resulting border/outline/shadow "too" thick and from memory the button component didn't add the extra border. I'm not sure which is the "correct" or preferred option.

As for the `inset` suggestion, I still believe this will make any box shadow we add begin at the controls edge and go inwards thus leaving the border from the `BorderControl`'s wrapper around the dropdown and `UnitControl` outside the box shadow. Hence the "double border".

The only way around that I saw was reworking the styling for all the inner components to the `BorderControl`.

Reply:
Very interesting that there isn't much consistency across the components, there, thanks for reviewing all those.

> As the current approach will address the "double borders" issue, I'm inclined to proceed with it, perhaps just with a tweak to make the border control dropdown button's hover style match its focus style.

Yes, that sounds good to me, particularly with the matching hover + focus style for the dropdown button. And with the shape of this PR, it looks like it should be pretty straightforward to adjust again in a follow-up if need be!

Reply:
Likewise, thanks for doing more digging!

> While making these changes this appeared to make the resulting border/outline/shadow "too" thick and from memory the button component didn't add the extra border. 

Sounds like it might be easier to go with your current approach then, particularly if it looks like it's going to be complex to get it working right.

Reply:
I took a run at refactoring the styles here to allow the `inset` focus box shadows and didn't hit as many edge cases as I thought. It did require a small negative margin hack to affect border collapsing but other than that came up well.

I'll resolve this conversation but wait for a final review.